extensions/s3/src/main/java/io/deephaven/extensions/s3/S3CompletableOutputStream.java

```diff
@@ -0,0 +1,374 @@
+import static io.deephaven.extensions.s3.S3Instructions.MIN_WRITE_PART_SIZE;
+
+class S3CompletableOutputStream extends CompletableOutputStream {
+
+    /**
+     * @see <a href="https://docs.aws.amazon.com/AmazonS3/latest/userguide/qfacts.html">Amazon S3 User Guide</a>
+     */
+    private static final int MIN_PART_NUMBER = 1;
+    private static final int MAX_PART_NUMBER = 10000;
+    private static final int INVALID_PART_NUMBER = -1;
+
+    private enum State {
+        OPEN, DONE, COMPLETED, ABORTED
+    }
+
+    private final S3Uri uri;
+    private final S3AsyncClient s3AsyncClient;
+    private final S3Instructions s3Instructions;
+
+    private final int writePartSize;
+    private final int numConcurrentWriteParts;
+
+    private final List<CompletedPart> completedParts;
+    private final List<OutgoingRequest> pendingRequests;
+
+    private int nextPartNumber;
+    private String uploadId; // Initialized on first write, changed back to null when multipart upload completed/aborted
+    private State state;
+
+    S3CompletableOutputStream(
+            @NotNull final URI uri,
+            @NotNull final S3AsyncClient s3AsyncClient,
+            @NotNull final S3Instructions s3Instructions) {
+        this.uri = s3AsyncClient.utilities().parseUri(uri);
+        this.s3AsyncClient = s3AsyncClient;
+        this.s3Instructions = s3Instructions;
+
+        this.writePartSize = s3Instructions.writePartSize();
+        this.numConcurrentWriteParts = s3Instructions.numConcurrentWriteParts();
+        this.pendingRequests = new ArrayList<>(numConcurrentWriteParts);
+
+        this.nextPartNumber = MIN_PART_NUMBER;
+        this.completedParts = new ArrayList<>();
+        this.state = State.OPEN;
+    }
+
+    @Override
+    public void write(final int b) throws IOException {
+        write((dest, destOff, destCount) -> {
+            verifyNotFull(dest);
+            dest.put((byte) b);
+            return 1;
+        }, 0, 1);
+    }
+
+    @Override
+    public void write(final byte @NotNull [] b) throws IOException {
+        write(b, 0, b.length);
+    }
+
+    @Override
+    public void write(final byte @NotNull [] b, final int off, final int len) throws IOException {
+        write((dest, currentOffset, remainingLength) -> {
+            verifyNotFull(dest);
+            final int lengthToWrite = Math.min(remainingLength, dest.remaining());
+            dest.put(b, currentOffset, lengthToWrite);
+            return lengthToWrite;
+        }, off, len);
+    }
+
+    @FunctionalInterface
+    private interface DataWriter {
+        /**
+         * Writes data to the given destination buffer, starting from the current offset in the source data.
+         *
+         * @param dest the destination buffer to write data to
+         * @param currentOffset the current offset in the source data
+         * @param remainingLength the remaining number of bytes of source data to write
```

Comment:
Does this include the total amount of remaining length for the source, or is this the maximum amount that dest can currently accept?

Reply:
This is the remaining length of source data.
We do a check for `final int lengthToWrite = Math.min(remainingLength, dest.remaining());` inside the DataWriter (see line 94).

Reply:
I think this means it should be a long, not an int? Otherwise, we are capped at 2GB?

Reply:
Not quite. It's the specific source (an array, or a byte). Int is entirely appropriate.

Reply:
I understand the source of the discrepancy, but I think matters are clearer now that this method and the interface are correctly marked private.